File: src/main/java/com/zotohlab/wflow/Composite.java
// This library is distributed in  the hope that it will be useful but without
// any  warranty; without  even  the  implied  warranty of  merchantability or
// fitness for a particular purpose.
// The use and distribution terms for this software are covered by the Eclipse
// Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
// can be found in the file epl-v10.html at the root of this distribution.
// By using this software in any  fashion, you are agreeing to be bound by the
// terms of this license. You  must not remove this notice, or any other, from
// this software.
// Copyright (c) 2013, Ken Leung. All rights reserved.

package com.zotohlab.wflow;

import java.util.ArrayList;
import java.util.List;
import java.util.Iterator;

/**
 * @author kenl
 *
 */
abstract class Composite extends Activity {

  private List<Activity> _innards= new ArrayList<>();

  public int size() { return _innards.size(); }

  protected Composite(String name) {
    super(name);
  }

  protected Composite() {}

  protected void add(Activity a) {
    _innards.add(a);
  }

  public Iterator<Activity> listChildren() {
    return _innards.listIterator();
  }

  public void realize(FlowNode fp) {
    CompositeNode p= (CompositeNode) fp;
    p.reifyInner( listChildren() );
  }

}


/**
 *
 * @author kenl
 *
 */
abstract class CompositeNode extends FlowNode {

  protected CompositeNode(FlowNode c, Activity a) {
    super(c,a);
  }

  public void reifyInner( Iterator<Activity> c) {
    _inner=new Innards(this,c);
  }

  public void reifyInner() {
    _inner=new Innards(this);
  }

  public Innards inner() { return _inner; }
  protected void setInner(Innards n) {
    _inner=n;
  }

  private Innards _inner = null;
}

/**
 *
 * @author kenl
 *
 */
class Innards {

  public boolean isEmpty() { return  _acts.size() == 0; }

  private List<Activity> _acts= new ArrayList<>();
  private FlowNode _outer;

  public Innards(FlowNode c, Iterator<Activity> a) {
    this(c);
    while (a.hasNext()) {
      _acts.add(a.next());
    }
  }

  public Innards(FlowNode outer) {
    _outer= outer;
  }

  public FlowNode next() {
    return _acts.size() > 0
      ? _acts.remove(0).reify(_outer)
    :
      null;
  }

  public int size() { return _acts.size(); }

}

File: src/main/clojure/czlab/xlib/dbio/composite.clj
;; This library is distributed in  the hope that it will be useful but without
;; any  warranty; without  even  the  implied  warranty of  merchantability or
;; fitness for a particular purpose.
;; The use and distribution terms for this software are covered by the Eclipse
;; Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
;; can be found in the file epl-v10.html at the root of this distribution.
;; By using this software in any  fashion, you are agreeing to be bound by the
;; terms of this license. You  must not remove this notice, or any other, from
;; this software.
;; Copyright (c) 2013-2015, Ken Leung. All rights reserved.

(ns ^{:doc ""
      :author "kenl" }

  czlab.xlib.dbio.composite

  (:require
    [czlab.xlib.util.core :refer [test-nonil try!]]
    [czlab.xlib.util.str :refer [hgl?]]
    [czlab.xlib.util.logging :as log])

  (:use [czlab.xlib.dbio.core]
        [czlab.xlib.dbio.sql])

  (:import
    [com.zotohlab.frwk.dbio Transactable SQLr MetaCache DBAPI]
    [java.sql Connection]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn CompositeSQLr*

  "A composite supports transactions"

  ^Transactable
  [^DBAPI db ]

  (reify

    Transactable

    (execWith [me func]
      (with-local-vars [rc nil]
      (with-open [conn
                  (.begin me) ]
        (try
          (->> (ReifySQLr
                 db
                 (fn [_] conn) #(%2 %1))
               (func )
               (var-set rc ))
          (.commit me conn)
          @rc
          (catch Throwable e#
            (.rollback me conn)
            (log/warn e# "")
            (throw e#))) )))

    (rollback [_ conn]
      (try! (-> ^Connection
                conn
                (.rollback))))

    (commit [_ conn]
      (-> ^Connection
          conn
          (.commit)))

    (begin [_]
      (let [conn (.open db) ]
        (.setAutoCommit conn false)
        (->> Connection/TRANSACTION_SERIALIZABLE
             (.setTransactionIsolation conn ))
        conn))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;EOF

